File: Q2.py
import math


def get_coordinates():
    coordinates = []
# loop until get x and y is done
    while True:
        x_str = input("Enter x coordinate (or 'done' to finish): ")

        if (x_str.lower()) == 'done':
            break

        y_str = input("Enter y coordinate: ")

        try:
            x = float(x_str)
            y = float(y_str)
            coordinates.append([x, y])
        except ValueError:
            print("Invalid input. Please enter valid numeric coordinates.")
# return the array
    return coordinates


def get_circles():
    coordinates = []
# loop until get x, y, radius is done
    while True:
        x_str = input("Enter x coordinate (or 'done' to finish): ")

        if (x_str.lower()) == 'done':
            break

        y_str = input("Enter y coordinate: ")
        radius_str = input("Enter radius: ")

        try:
            x = float(x_str)
            y = float(y_str)
            radius = float(radius_str)
            coordinates.append([x, y, radius])
        except ValueError:
            print("Invalid input. Please enter valid numeric coordinates.")
# return the array
    return coordinates


def is_inside_circle(point, circle):
    # Calculate the distance between the point and the center of the circle
    x, y, r = circle
    # Euclidean distance between the given point and the center of the circle using the Pythagorean theorem.
    distance = math.sqrt((point[0] - x) ** 2 + (point[1] - y) ** 2)

    # Check if the point is inside the circle (including the border) (True or False)
    return distance <= r


def points_inside_circles(points, circles):
    result = []
    # loop in for each circle to find all point
    for circle in circles:
        points_inside = []
        # all the points per circle
        for point in points:
            if is_inside_circle(point, circle):
                points_inside.append(point)
        result.append(points_inside)

    return result


# get input for points
print("\nplease input coordinates for points\n")
points = get_coordinates()
# get input for circles
print("\nplease input coordinates for circles\n")
circles = get_circles()

# put both arrays into the function to find the result
result = points_inside_circles(points, circles)

# print all the results
for i, points_inside_circle in enumerate(result):
    print(f"Points inside Circle {i + 1}: {points_inside_circle}")

File: Q2.js
/**********************************************************
this is basically a javascript version of Q2 code
But prompt is not vanilla JS so I originally wrote
in PY (not my favoriate). but anyways P'Patr said
pure JS so I rewrote it

*** Note ****
Please install npm prompt-sync first

By: Paul J.
***********************************************************/

const prompt = require("prompt-sync")();


/************************************************************************
 this function is for getting the coordinates or the point, this will
 catch error. 
 ************************************************************************/
function getCoordinates() {
    const coordinates = [];

    while (true) {
        const xStr = prompt("Enter x coordinate (or 'done' to finish):");

        if (xStr.toLowerCase() === 'done') {
            break;
        }

        const yStr = prompt("Enter y coordinate:");

        try {
            const x = parseFloat(xStr);
            const y = parseFloat(yStr);
            coordinates.push([x, y]);
        //catch error and tell user
        } catch (error) {
            console.log("Invalid input. Please enter valid numeric coordinates.");
        }
    }

    return coordinates;
}

/************************************************************************
 similar function to the top but for getting the x y and radius of circle
 ************************************************************************/
function getCircles() {
    const circles = [];

    while (true) {
        const xStr = prompt("Enter x coordinate (or 'done' to finish):");

        if (xStr.toLowerCase() === 'done') {
            break;
        }

        const yStr = prompt("Enter y coordinate:");
        const radiusStr = prompt("Enter radius:");

        try {
            const x = parseFloat(xStr);
            const y = parseFloat(yStr);
            const radius = parseFloat(radiusStr);
            circles.push([x, y, radius]);
        } catch (error) {
            console.log("Invalid input. Please enter valid numeric coordinates.");
        }
    }

    return circles;
}

/************************************************************************
 this function is for checking whether point is in the circle or not
 use for the function below
 ************************************************************************/
function isInsideCircle(point, circle) {
    const [x, y, r] = circle;
    //Euclidean distance between the given point and the center of the circle using the Pythagorean theorem.
    const distance = Math.sqrt((point[0] - x) ** 2 + (point[1] - y) ** 2);
    return distance <= r;
}

/************************************************************************
 this function is for looping and check for circle in array of circles
 and while we are at that circle we check for the point in the array of
 points using the function above
 ************************************************************************/
function pointsInsideCircles(points, circles) {
    const result = [];

    for (const circle of circles) {
        const pointsInside = [];

        for (const point of points) {
            if (isInsideCircle(point, circle)) {
                pointsInside.push(point);
            }
        }

        result.push(pointsInside);
    }

    return result;
}


//ask for points
console.log("\nPlease input coordinates for points:\n");
const points = getCoordinates();
//ask for cirlces
console.log("\nPlease input coordinates for circles:\n");
const circles = getCircles();
//put those points and circle into this function to return a array
const result = pointsInsideCircles(points, circles);
//here in JS we cannot print object, the returned value is a JSON so we need to make it a string first
for (let i = 0; i < result.length; i++) {
    console.log(`Points inside Circle ${i + 1}: ${JSON.stringify(result[i])}`);
}
